python/pywatchman/aioclient.py

```diff
@@ -0,0 +1,340 @@
+
+from . import compat, encoding
+
+from . import CommandError, SocketTimeout, WatchmanError
+
+
+# 2 bytes marker, 1 byte int size, 8 bytes int64 value
+SNIFF_LEN = 13
+
+
+# TODO: Fix this when https://github.com/python/asyncio/issues/281 is resolved.
+# tl;dr is that you cannot have different event loops running in different
+# threads all fork subprocesses and listen for child events. The current
+# workaround is to do the old fashioned blocking process communication using a
+# ThreadPool.
+def _resolve_sockname_helper():
+    # if invoked via a trigger, watchman will set this env var; we
+    # should use it unless explicitly set otherwise
+    path = os.getenv('WATCHMAN_SOCK')
+    if path:
+        return path
+
+    cmd = ['watchman', '--output-encoding=bser', 'get-sockname']
+
+    try:
+        p = subprocess.Popen(cmd,
+                             stdout=subprocess.PIPE,
+                             stderr=subprocess.PIPE,
+                             close_fds=os.name != 'nt')
+    except OSError as e:
+        raise WatchmanError('"watchman" executable not in PATH (%s)', e)
+
+    stdout, stderr = p.communicate()
+    exitcode = p.poll()
+
+    if exitcode:
+        raise WatchmanError('watchman exited with code %d' % exitcode)
+
+    result = bser.loads(stdout)
+
+    if 'error' in result:
+        raise WatchmanError(str(result['error']))
+
+    return result['sockname']
+
+
+async def _resolve_sockname():
+    '''Find the Unix socket path to the global Watchman instance.'''
+    loop = asyncio.get_event_loop()
+    return await loop.run_in_executor(None, _resolve_sockname_helper)
+
+
+class AsyncTransport(object):
+    '''Communication transport to the Watchman Service.'''
+
+    async def activate(self, **kwargs):
+        '''Make the transport ready for use. Optional for subclasses.'''
+        pass
+
+    async def read(self, size):
+        '''Read 'size' bytes from the transport.'''
+        raise NotImplementedError()
+
+    async def write(self, buf):
+        '''Write 'buf' bytes to the transport.'''
+        raise NotImplementedError()
+
+    def close(self):
+        '''Close the transport. Optional for subclasses.'''
+        pass
+
+
+class AsyncUnixSocketTransport(AsyncTransport):
+    '''Local Unix domain socket transport supporting asyncio.'''
+
+    def __init__(self):
+        self.sockname = None
+        self.reader = None
+        self.writer = None
+
+    async def activate(self, **kwargs):
+        # Requires keyword-argument 'sockname'
+        reader, writer = await asyncio.open_unix_connection(kwargs['sockname'])
+        self.reader = reader
+        self.writer = writer
+
+    async def write(self, data):
+        self.writer.write(data)
+        await self.writer.drain()
+
+    async def read(self, size):
+        res = await self.reader.read(size)
+        return res
+
+    def close(self):
+        if self.writer:
+            self.writer.close()
+
+
+class AsyncCodec(object):
```

Comment:
nit: leave this as-is but in the future keep in mind that in Python "codec" is a protocol that contains the methods `encode` and `decode`. `send`/`receive` implies connections, most often sockets. Funnily, you call the instantiated AsyncBserCodec in line 284 a `connection` yourself :)